### DISCUSSION TITLE
running Rust tests from within helix

### DISCUSSION THREAD

**Discussion by blitzerr** [helix-editor/helix#5014] [General] [answered]
How do people run rust tests from within the helix editor ?

Like in Vscode, you have a little symbol over the functions marked as `#[test]` and you can run it with a click of a button. Is there a similar way in helix to run the test under the cursor with a key-press and display the success or failures in the status bar?

**Comment by the-mikedavis** [comment, marked as answer]
Running tests from Helix is not currently possible. Generally this is something provided by plugins in other editors. https://build-server-protocol.github.io/ might be a way to provide this generically in the future but it is too immature now. For now, you can run build and test tools in separate terminal sessions.

**Comment by blitzerr** [comment]
Thanks a lot @the-mikedavis

**Comment by sebashwa** [comment]
Hey, I miss running my tests from within helix as well.
@the-mikedavis build-server-protocol sounds great, is there any way to already contribute to push this topic?

For me as a workaround (as soon as https://github.com/helix-editor/helix/issues/3134 is possible) I thought of something like:

* Have a small shell script in your project dir which runs tests and writes the output to a file e.g. `.test-results`
* Use a shortcut to open a split pane in tmux with `tail --follow .test-results`
* Use a helix shortcut to run the script with current file / line as args

I know that not every test runner can deal with lines to run single tests though (pytest for example needs class / function names).
Is it feasible to also get this context information (maybe from tree sitter) when running `sh` from helix?

**Reply by the-mikedavis** [reply]
> build-server-protocol sounds great, is there any way to already contribute to push this topic?

Not really no, it's too immature. There are only a handful of servers and AFAIK only one editor that implements the client side of it.

Current class/function information could probably be gathered from the textobjects and the current primary cursor position (see https://github.com/helix-editor/helix/pull/3944)

**Comment by dmyyy** [comment]
I wrote a small shell script that runs all unit tests in your current buffer that can work in the interim.

```
[keys.normal.space.t]
t = [
  ":sh RUSTFLAGS=-Awarnings cargo test -q -- --nocapture %sh{echo %{buffer_name} | sed -E 's|^src/||; s|/mod\\.rs$||; s|\\.rs$||; s|/|::|g' }",
]
```
